airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/dataflow/state/StateKeyClient.kt

```diff
@@ -51,57 +51,64 @@ class SelfDescribingStateKeyClient : StateKeyClient {
  */
 @Singleton
 @Requires(property = "airbyte.destination.core.data-channel.medium", value = "STDIO")
-class InferredStateKeyClient(private val catalog: DestinationCatalog) : StateKeyClient {
-    // sequence of all state messages
-    private val globalCounter = AtomicLong(1)
+class InferredStateKeyClient(
+    private val catalog: DestinationCatalog,
+) : StateKeyClient {
 
-    // sequence of state messages per stream
-    private var streamCounters = ConcurrentHashMap<DestinationStream.Descriptor, AtomicLong>()
+    private val counters = ConcurrentHashMap<DestinationStream.Descriptor, AtomicLong>()
 
-    // caches keys / strings for perf reasons
     private val keyCache = ConcurrentHashMap<DestinationStream.Descriptor, PartitionKey>()
 
     override fun getPartitionKey(msg: DestinationRecordRaw): PartitionKey =
-        getCachedKeyForDesc(msg.stream.unmappedDescriptor)
-
-    override fun getStateKey(msg: CheckpointMessage): StateKey {
-        val ordinal = globalCounter.getAndIncrement()
+        currentKey(msg.stream.unmappedDescriptor)
 
+    override fun getStateKey(msg: CheckpointMessage): StateKey =
         when (msg) {
             is StreamCheckpoint -> {
-                val partitions = listOf(getCachedKeyForDesc(msg.checkpoint.unmappedDescriptor))
-                incrementForDesc(msg.checkpoint.unmappedDescriptor)
-                return StateKey(ordinal, partitions)
+                val desc = msg.checkpoint.unmappedDescriptor
+
+                val ordinal = currentOrdinal(desc)
+                val partitions = listOf(currentKey(desc))
+                advance(desc)
+                StateKey(ordinal, partitions)
             }
             is GlobalCheckpoint,
             is GlobalSnapshotCheckpoint -> {
-                val partitions =
-                    catalog.streams.map {
-                        val key = getCachedKeyForDesc(it.unmappedDescriptor)
-                        incrementForDesc(it.unmappedDescriptor)
-                        key
-                    }
-                return StateKey(ordinal, partitions)
+                val streams = catalog.streams
+                require(streams.isNotEmpty()) {
```

Comment:
happy to keep this check for clarity, but fyi we do validate this in the DestinationCatalog constructor https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/command/DestinationCatalog.kt#L34